### ISSUE TITLE
K8SSAND-1563 ⁃ error "Exception encountered during startup: The seed provider lists no seeds" when recreating pods

### ISSUE THREAD

**Issue by Tom910** [k8ssandra/k8ssandra-operator#567]
**What happened?**

When I have many cass clusters and run `kubectl delete --all pods --namespace k8s-cluster-test` after this command 70% of clusters would recover but 30% of clusters would have the error `Exception encountered during startup: The seed provider lists no seeds.`

cluster logs

```
WARN  [main] 2022-06-14 11:52:25,312 K8SeedProvider4x.java:58 - Seed provider couldn't lookup host test-clusterr-37-seed-service
WARN  [main] 2022-06-14 11:52:25,338 K8SeedProvider4x.java:58 - Seed provider couldn't lookup host test-clusterr-37-dctestr-37-additional-seed-service
ERROR [main] 2022-06-14 11:52:25,340 CassandraDaemon.java:909 - Exception encountered during startup: The seed provider lists no seeds.
```

dns check on cass nodes

```
❯ kubectl exec --stdin --tty -n k8s-cluster-test test-clusterr-37-dctestr-37-default-sts-0 -- /bin/bash

cassandra@test-clusterr-37-dctestr-37-default-sts-0:/$ curl test-clusterr-37-seed-service
curl: (6) Could not resolve host: test-clusterr-37-seed-service

cassandra@test-clusterr-37-dctestr-37-default-sts-0:/$ curl test-clusterr-37-dctestr-37-all-pods-service
curl: (7) Failed to connect to test-clusterr-37-dctestr-37-all-pods-service port 80: Connection refused
```

At the same time, the cluster is repaired if you delete the pods again

**Did you expect to see something different?**

I want to see 100% recover clusters

**How to reproduce it (as minimally and precisely as possible)**:

**Environment**

* K8ssandra Operator version:

    `github.com/k8ssandra/k8ssandra-operator/config/deployments/control-plane/cluster-scope?ref=v1.1.1`
    
    * Kubernetes version information:

    `v1.21.5` coredns = 1.6.3
    * Kubernetes cluster kind:

```
managed
```

* Manifests:

I have 40 similar clusters like:

```
apiVersion: k8ssandra.io/v1alpha1
kind: K8ssandraCluster
metadata:
  name: test-clusterr-37
  namespace: k8s-cluster-test
spec:
  cassandra:
    serverVersion: "4.0.1"
    softPodAntiAffinity: true
    resources:
      requests:
        cpu: "500m"
        memory: "1Gi"
      limits:
        cpu: "1900m"
        memory: "2Gi"
    datacenters:
      - metadata:
          name: dctestr-37
        size: 2
        storageConfig:
          cassandraDataVolumeClaimSpec:
            resources:
              requests:
                storage: 2Gi
        config:
          jvmOptions:
            heapSize: 256M

```

* K8ssandra Operator Logs:

```
1.655210188848521e+09   INFO    controller.k8ssandracluster     Reconciling Medusa user secrets {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test-clusterr-37", "namespace": "k8s-cluster-test", "K8ssandraCluster": "k8s-cluster-test/test-clusterr-37"}
1.6552101888485692e+09  INFO    controller.k8ssandracluster     Medusa user secrets successfully reconciled     {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test-clusterr-37", "namespace": "k8s-cluster-test", "K8ssandraCluster": "k8s-cluster-test/test-clusterr-37"}
1.655210188848572e+09   INFO    controller.k8ssandracluster     Reconciling replicated secrets  {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test-clusterr-37", "namespace": "k8s-cluster-test", "K8ssandraCluster": "k8s-cluster-test/test-clusterr-37"}
1.655210188848788e+09   INFO    controller.k8ssandracluster     Medusa reconcile for dctestr-37 on namespace k8s-cluster-test   {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test-clusterr-37", "namespace": "k8s-cluster-test", "K8ssandraCluster": "k8s-cluster-test/test-clusterr-37"}
1.655210188848799e+09   INFO    controller.k8ssandracluster     Medusa is not enabled   {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test-clusterr-37", "namespace": "k8s-cluster-test", "K8ssandraCluster": "k8s-cluster-test/test-clusterr-37"}
1.6552101888513434e+09  INFO    controller.k8ssandracluster     Reconciling seeds       {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test-clusterr-37", "namespace": "k8s-cluster-test", "K8ssandraCluster": "k8s-cluster-test/test-clusterr-37", "CassandraDatacenter": "k8s-cluster-test/dctestr-37", "K8SContext": ""}
1.6552101888514278e+09  INFO    controller.k8ssandracluster     The datacenter is reconciled    {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test-clusterr-37", "namespace": "k8s-cluster-test", "K8ssandraCluster": "k8s-cluster-test/test-clusterr-37", "CassandraDatacenter": "k8s-cluster-test/dctestr-37", "K8SContext": ""}
1.6552101888591702e+09  DEBUG   controller-runtime.webhook.webhooks     received request        {"webhook": "/validate-k8ssandra-io-v1alpha1-k8ssandracluster", "UID": "bd0f9a93-61da-44ff-88fb-3f19a6955a6c", "kind": "k8ssandra.io/v1alpha1, Kind=K8ssandraCluster", "resource": {"group":"k8ssandra.io","version":"v1alpha1","resource":"k8ssandraclusters"}}
1.6552101888596108e+09  INFO    k8ssandracluster-webhook        validate K8ssandraCluster update        {"K8ssandraCluster": "test-clusterr-37"}
1.6552101888596413e+09  DEBUG   controller-runtime.webhook.webhooks     wrote response  {"webhook": "/validate-k8ssandra-io-v1alpha1-k8ssandracluster", "code": 200, "reason": "", "UID": "bd0f9a93-61da-44ff-88fb-3f19a6955a6c", "allowed": true}
1.6552101888640745e+09  INFO    controller.k8ssandracluster     Preparing to update replication for system keyspaces    {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test-clusterr-37", "namespace": "k8s-cluster-test", "K8ssandraCluster": "k8s-cluster-test/test-clusterr-37", "CassandraDatacenter": "k8s-cluster-test/dctestr-37", "K8SContext": "", "replication": {"dctestr-37":2}}
1.6552101888641243e+09  INFO    controller.k8ssandracluster     Ensuring that keyspace system_traces exists in cluster test-clusterr-37...      {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test-clusterr-37", "namespace": "k8s-cluster-test", "K8ssandraCluster": "k8s-cluster-test/test-clusterr-37", "CassandraDatacenter": "k8s-cluster-test/dctestr-37", "K8SContext": ""}
1.6552101888643324e+09  ERROR   controller.k8ssandracluster     Failed to fetch datacenter pods {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test-clusterr-37", "namespace": "k8s-cluster-test", "K8ssandraCluster": "k8s-cluster-test/test-clusterr-37", "CassandraDatacenter": "k8s-cluster-test/dctestr-37", "K8SContext": "", "error": "no pods in READY state found in datacenter dctestr-37"}
github.com/k8ssandra/k8ssandra-operator/pkg/cassandra.(*defaultManagementApiFacade).EnsureKeyspaceReplication
        /workspace/pkg/cassandra/management.go:289
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).updateReplicationOfSystemKeyspaces
        /workspace/controllers/k8ssandra/schemas.go:183
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).checkSchemas
        /workspace/controllers/k8ssandra/schemas.go:43
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).reconcileDatacenters
        /workspace/controllers/k8ssandra/datacenters.go:168
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).reconcile
        /workspace/controllers/k8ssandra/k8ssandracluster_controller.go:133
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).Reconcile
        /workspace/controllers/k8ssandra/k8ssandracluster_controller.go:87
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:114
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:311
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227
1.6552101888643928e+09  ERROR   controller.k8ssandracluster     Failed to update replication    {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test-clusterr-37", "namespace": "k8s-cluster-test", "K8ssandraCluster": "k8s-cluster-test/test-clusterr-37", "CassandraDatacenter": "k8s-cluster-test/dctestr-37", "K8SContext": "", "keyspace": "system_traces", "error": "no pods in READY state found in datacenter dctestr-37"}
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).checkSchemas
        /workspace/controllers/k8ssandra/schemas.go:43
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).reconcileDatacenters
        /workspace/controllers/k8ssandra/datacenters.go:168
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).reconcile
        /workspace/controllers/k8ssandra/k8ssandracluster_controller.go:133
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).Reconcile
        /workspace/controllers/k8ssandra/k8ssandracluster_controller.go:87
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:114
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:311
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227
1.6552101888729985e+09  INFO    controller.k8ssandracluster     updated k8ssandracluster status {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test-clusterr-37", "namespace": "k8s-cluster-test", "K8ssandraCluster": "k8s-cluster-test/test-clusterr-37"}
1.6552101888730514e+09  ERROR   controller.k8ssandracluster     Reconciler error        {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test-clusterr-37", "namespace": "k8s-cluster-test", "error": "no pods in READY state found in datacenter dctestr-37"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227
```

**Anything else we need to know?**:

I found a similar issue - https://github.com/kubernetes/kubernetes/issues/92559 about DNS cache but I don't have oportunity change this parameters right now

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-1563) by [Unito](https://www.unito.io)
┆friendlyId: K8SSAND-1563
┆priority: Medium

**Comment by Tom910** [comment]
Additional information
```
❯ kubectl get pods test-clusterr-37-dctestr-37-default-sts-0 -n k8s-cluster-test -o yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: "2022-06-14T11:21:21Z"
  generateName: test-clusterr-37-dctestr-37-default-sts-
  labels:
    app.kubernetes.io/instance: cassandra-test-clusterr-37
    app.kubernetes.io/managed-by: cass-operator
    app.kubernetes.io/name: cassandra
    app.kubernetes.io/version: 4.0.1
    cassandra.datastax.com/cluster: test-clusterr-37
    cassandra.datastax.com/datacenter: dctestr-37
    cassandra.datastax.com/node-state: Starting
    cassandra.datastax.com/rack: default
    controller-revision-hash: test-clusterr-37-dctestr-37-default-sts-78698c559b
    statefulset.kubernetes.io/pod-name: test-clusterr-37-dctestr-37-default-sts-0
  name: test-clusterr-37-dctestr-37-default-sts-0
```

```
❯ kubectl get svc test-clusterr-37-seed-service -n k8s-cluster-test -o yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    cassandra.datastax.com/resource-hash: n6slxhCbAHOHCzjD4a6VkgBOeGLkAVQ+cLrpmZHxNKI=
  creationTimestamp: "2022-06-10T11:55:32Z"
  labels:
    app.kubernetes.io/instance: cassandra-test-clusterr-37
    app.kubernetes.io/managed-by: cass-operator
    app.kubernetes.io/name: cassandra
    app.kubernetes.io/version: 4.0.1
    cassandra.datastax.com/cluster: test-clusterr-37
  name: test-clusterr-37-seed-service
  namespace: k8s-cluster-test
  ownerReferences:
  - apiVersion: cassandra.datastax.com/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: CassandraDatacenter
    name: dctestr-37
    uid: ed346db6-6df8-47e3-b555-0a55139f92c2
  resourceVersion: "397321"
  uid: 0ca4f1a1-0f22-47e5-a892-42d7304af523
spec:
  clusterIP: None
  clusterIPs:
  - None
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  publishNotReadyAddresses: true
  selector:
    cassandra.datastax.com/cluster: test-clusterr-37
    cassandra.datastax.com/seed-node: "true"
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
```

as result pod didn't have `cassandra.datastax.com/seed-node: "true"` and service don't work correctly

**Comment by jsanda** [comment]
This doesn't answer the question about what's happening, but I want to ask what is the use case for deleting the pods? Have you considered using the `stopped` property to shutdown DCs? That will scale the underlying statefulsets down to zero but keep the persistent volumes intact.

**Comment by jsanda** [comment]
This should be investigated. At any point for various reasons, Kubernetes can delete and reschedule pods. k8ssandra-operator and cass-operator need to handle that gracefully. 

It's also worth noting that cass-operator creates a PodDisruptionBudget for each CassadraDatacenter. It only sets the `minAvailable` property to a value of the datacenter size minus one.

**Comment by Tom910** [comment]
> This should be investigated. At any point for various reasons, Kubernetes can delete and reschedule pods. k8ssandra-operator and cass-operator need to handle that gracefully.

In my case it was cluster for test and I used preemptible workers which are automatically deleted

**Comment by hvintus** [comment]
The problem seems to be that `startOneNodePerRack` starts the first node without seed label (since it assumes there still are other seeds), but by the time Cassandra daemon is started, the seed services have no endpoints anymore. `cass-operator` doesn't handle that case in further reconciliation attempts stopping at `findStartingNodes`.

The reason there still are IPs in the seed endpoints when all the pods have been deleted is the `k8ssanda-operator` behaviour. It gathers all the pods with seed labels from all the clusters and posts them in the `-addional-seed-service`. In case of the single cluster deployment it effectively copies IPs from the `-seed-service` to the `-additional-seed-service`, just with a delay caused by its reconciliation loop. Since `cass-operator` considers both services when deciding to assign the seed label to the starting node, that pod ends up without the label.

**Comment by jsanda** [comment]
>The problem seems to be that `startOneNodePerRack` starts the first node without seed label (since it assumes there still are other seeds)

This is incorrect. Here is the [snippet](https://github.com/k8ssandra/cass-operator/blob/34648fd39f11ae76b42a9aaea6d283a3861ddc40/pkg/reconciliation/reconcile_racks.go#L1832-L1832) from the method:

```go
			if labelSeedBeforeStart {
					patch := client.MergeFrom(pod.DeepCopy())
					pod.Labels[api.SeedNodeLabel] = "true"
					if err := rc.Client.Patch(rc.Ctx, pod, patch); err != nil {
						return "", err
					}

					rc.Recorder.Eventf(rc.Datacenter, corev1.EventTypeNormal, events.LabeledPodAsSeed,
						"Labeled pod a seed node %s", pod.Name)

					// sleeping five seconds for DNS paranoia
					time.Sleep(5 * time.Second)
				}
```
You are correct that for a single-DC cluster the seed-service and additional-seeds-service will the same endpoints. When the pods get deleted, those endpoints will get removed from the seed-service. The next time k8ssandra-operators performs reconciliation after the endpoints are removed from the seed-service it will delete the endpoints for the additional-seeds-service. At this point cass-operator should proceed with starting a node in each rack.

There is obviously some sort of timing issue since @Tom910 mentioned this didn't happen all the time and deleting the pods a second time got things back to a healthy state. 

For what it's worth, I've had trouble with preemptible worker nodes and cass-operator in the past and avoided them because of it. 

This is going to require some more investigation.

**Comment by jsanda** [comment]
Please [add your planning poker estimate](https://app.zenhub.com/workspaces/K8ssandra-Grooming-623b197f1fcd4800110540ef/issues/k8ssandra/k8ssandra-operator/567?planning-poker) with ZenHub @burmanm

**Comment by hvintus** [comment]
@jsanda You are right that `startOneNodePerRack` could label the pod as seed, but the problem is that it won't in that particular case:

```go
	externalSeedPoints := 0
	if existingEndpoints, err := rc.GetAdditionalSeedEndpoint(); err == nil {
		externalSeedPoints = len(existingEndpoints.Subsets[0].Addresses)
	}

	labelSeedBeforeStart := readySeeds == 0 && len(rc.Datacenter.Spec.AdditionalSeeds) == 0 && externalSeedPoints == 0
```

`externalSeedPoints` is not 0 since there still are IP address in the `-addional-seed-service` endpoints. The sequence of events I observer when I delete all the worker pods:

1. Before termination:
	* pod1: Started, Seed
	* pod2: Started, Seed
	* pod3: Started, Seed
	* seed-service: [Pod1, Pod2, Pod3]
	* additional-seed-service: [Pod1, Pod2, Pod3]
2. Pods terminated:
	* seed-service: []
	* additional-seed-service: [Pod1, Pod2, Pod3]
3. Pods recreated: 
	* pod1: Ready-to-Start
	* pod2: Ready-to-Start
	* pod3: Ready-to-Start
	* seed-service: []
	* additional-seed-service: [Pod1, Pod2, Pod3]
4. First pod is being started: 
	* pod1: Starting
	* pod2: Ready-to-Start
	* pod3: Ready-to-Start
	* seed-service: []
	* additional-seed-service: [Pod1, Pod2, Pod3]
4. Additional seed service refreshed: 
	* pod1: Starting
	* pod2: Ready-to-Start
	* pod3: Ready-to-Start
	* seed-service: []
	* additional-seed-service: []
	
One of the fixes I see here is to exclude endpoints gathered from the DC when refreshing its `-additional-seed-service`. That way they would be truly *additional*.

`k8ssandra-operator/controllers/k8ssandra/seeds.go:newEndpoints`:
```go
	for _, seed := range seeds {
		// When building endpoints for `dc`, exclude pods gathered from the `dc` itself.
		if labels.GetLabel(&seed, cassdcapi.DatacenterLabel) != dc.Name {
			addresses = append(addresses, corev1.EndpointAddress{
				IP: seed.Status.PodIP,
			})
		}
	}
```

With the fix the sequence becomes such:
3. Pods recreated: 
	* pod1: Ready-to-Start
	* pod2: Ready-to-Start
	* pod3: Ready-to-Start
	* seed-service: []
	* additional-seed-service: **[]**
4. First pod is being started: 
	* pod1: Starting, **Seed**
	* pod2: Ready-to-Start
	* pod3: Ready-to-Start
	* seed-service: [**Pod1**]
	* additional-seed-service: **[]**
4. Additional seed service refreshed: 
	* pod1: Starting, Seed
	* pod2: Ready-to-Start
	* pod3: Ready-to-Start
	* seed-service: [**Pod1**]
	* additional-seed-service: **[]**
	
Additional changes are required in order for the clusters to become stuck at start-up without seeds in more general case. One way to do that would be to change `findStartingNodes` to restore Seed label on the starting pod if no other seeds are discovered at that phase.

**Comment by jsanda** [comment]
As I mentioned in my previous comment, I think there may either be a timing issue or something that fails to trigger reconciliation in k8ssandra-operator sometimes. I say since the error doesn't happen all the time. 

I have tried to reproduce several times locally with a kind cluster and have been unable to do so. I deployed a 3 node Cassandra cluster. Every time the additional seeds service endpoints get deleted eventually. The pods get recreated and return to the ready state.

We could change the logic in seeds.go as you described, but I would still like to determine the root cause. 

@hvintus how are you reproducing this?
